src/core/ad.jl

```diff
@@ -1,83 +1,120 @@
+
+    # Define function to compute log joint.
+    function f(θ)
+        vi[sampler] = θ
+        return -runmodel(model, vi, sampler).logp
+    end
 
-  # Define function to compute log joint.
-  function f(θ)
-      vi[spl] = θ
-      return -runmodel(model, vi, spl).logp
-  end
+    # Set chunk size and do ForwardMode.
+    chunk = ForwardDiff.Chunk(min(length(θ), chunk_size))
+    config = ForwardDiff.GradientConfig(f, θ, chunk)
+    ∂l∂θ = ForwardDiff.gradient!(similar(θ), f, θ, config)
+    l = vi.logp.value
 
-  # Set chunk size and do ForwardMode.
-  chunk = ForwardDiff.Chunk(min(length(θ), chunk_size))
-  config = ForwardDiff.GradientConfig(f, θ, chunk)
-  ∂l∂θ = ForwardDiff.gradient!(similar(θ), f, θ, config)
+    # Replace old parameters to ensure this function doesn't mutate `vi`.
+    vi.vals, vi.logp = vals_old, logp_old
 
-  # Replace old parameters to ensure this function doesn't mutate `vi`.
-  vi[spl] = θ_old
+    # Strip tracking info from θ to avoid mutating it.
+    θ .= ForwardDiff.value.(θ)
 
-  return getlogp(vi), ∂l∂θ
+    return l, ∂l∂θ
 end
 
 """
 gradient_reverse(
-  θ::AbstractVector{<:Real},
-  vi::VarInfo,
-  model::Function,
-  spl::Union{Nothing, Sampler}=nothing,
+    θ::AbstractVector{<:Real},
+    vi::VarInfo,
+    model::Function,
+    sampler::Union{Nothing, Sampler}=nothing,
 )
 
-Computes the gradient of the log joint of `θ` for the model specified by `(vi, spl, model)`
-using reverse-mode AD from Flux.jl.
+Computes the gradient of the log joint of `θ` for the model specified by
+`(vi, sampler, model)` using reverse-mode AD from Flux.jl.
 """
 function gradient_reverse(
-  θ::AbstractVector{<:Real},
-  vi::Turing.VarInfo,
-  model::Function,
-  spl::Union{Nothing, Sampler}=nothing,
+    θ::AbstractVector{<:Real},
+    vi::Turing.VarInfo,
+    model::Function,
+    sampler::Union{Nothing, Sampler}=nothing,
 )
-  # Specify objective function.
-  function f(θ)
-    vi[spl] = θ
-    return -runmodel(model, vi, spl).logp
-  end
-
-  # Compute forward and reverse passes.
-  l, ȳ = Tracker.forward(f, θ)
-  ∂l∂θ = ȳ(1)[1]
-
-  # Remove tracking info from variables in model (because mutable state).
-  vi.logp = Tracker.data(vi.logp)
-  vi[spl] .= Tracker.data.(vi[spl])
-
-  # Return non-tracked gradient value
-  return Tracker.data(l), Tracker.data(∂l∂θ)
+    vals_old, logp_old = copy(vi.vals), copy(vi.logp)
+
+    # Specify objective function.
+    function f(θ)
+        vi[sampler] = θ
+        return -runmodel(model, vi, sampler).logp
+    end
+
+    # Compute forward and reverse passes.
+    l_tracked, ȳ = Tracker.forward(f, θ)
+    l, ∂l∂θ = Tracker.data(l_tracked), Tracker.data(ȳ(1)[1])
+
+    # Remove tracking info from variables in model (because mutable state).
+    vi.vals, vi.logp = vals_old, logp_old
+
+    # Strip tracking info from θ to avoid mutating it.
+    θ .= Tracker.data.(θ)
+
+    # Return non-tracked gradient value
+    return l, ∂l∂θ
```

Comment:
Do you think it might be good for us to wrapper the result with `DiffResults.DiffResult` (http://www.juliadiff.org/DiffResults.jl/stable/#Constructing-a-DiffResult-1). And retrieve them later using `DiffResults.value` and `DiffResults.gradient`.

Reply:
Good idea, but let's do this in a separate PR when incorporating DynamicHMC.